src/plugins/upf/upf.c

```diff
@@ -773,6 +773,380 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
+  u8 *name = NULL;
+  clib_error_t *error = NULL;
+  int rv = 0;
+
+  /* Get a line of input. */
+  if (!unformat_user (input, unformat_line_input, line_input))
+    return error;
+
+  while (unformat_check_input (line_input) != UNFORMAT_END_OF_INPUT)
+    {
+      if (unformat (line_input, "%s", &name))
+        break;
+      else
+        {
+          error = unformat_parse_error (line_input);
+          goto done;
+        }
+    }
+
+  rv = vnet_upf_app_add_del(name, 0);
+
+  switch (rv)
+    {
+    case 0:
+      break;
+
+    case VNET_API_ERROR_VALUE_EXIST:
+      error = clib_error_return (0, "application already exists...");
+      break;
+
+    case VNET_API_ERROR_NO_SUCH_ENTRY:
+      error = clib_error_return (0, "application does not exist...");
+      break;
+
+    default:
+      error = clib_error_return (0, "%s returned %d", __FUNCTION__, rv);
+      break;
+    }
+
+done:
+  vec_free (name);
+  unformat_free (line_input);
+
+  return error;
+}
+
+/* *INDENT-OFF* */
+VLIB_CLI_COMMAND (upf_delete_app_command, static) =
+{
+  .path = "delete upf application",
+  .short_help = "delete upf application <name>",
+  .function = upf_delete_app_command_fn,
+};
+/* *INDENT-ON* */
+
+static int
+vnet_upf_rule_add_del(u8 * app_name, u32 rule_index, u8 add)
+{
+  upf_main_t *sm = &upf_main;
+  uword *p = NULL;
+  upf_dpi_app_t *app = NULL;
+  upf_dpi_rule_t *rule = NULL;
+
+  p = hash_get_mem (sm->upf_app_by_name, app_name);
+  if (!p)
+    return VNET_API_ERROR_NO_SUCH_ENTRY;
+
+  app = pool_elt_at_index (sm->upf_apps, p[0]);
+
+  p = hash_get_mem (app->rules_by_id, &rule_index);
+
+  if (add)
+    {
+      if (p)
+        return VNET_API_ERROR_VALUE_EXIST;
+
+      pool_get (app->rules, rule);
+      memset(rule, 0, sizeof(*rule));
+      rule->id = rule_index;
+
+      hash_set_mem (app->rules_by_id,
+                    &rule_index, rule - app->rules);
+    }
+  else
+    {
+      if (!p)
+        return VNET_API_ERROR_NO_SUCH_ENTRY;
+
+      rule = pool_elt_at_index (app->rules, p[0]);
+      hash_unset_mem (app->rules_by_id, &rule_index);
+      pool_put (app->rules, rule);
+    }
+
+  return 0;
+}
+
+static clib_error_t *
+upf_application_rule_add_del_command_fn (vlib_main_t * vm,
+                                         unformat_input_t * input,
+                                         vlib_cli_command_t * cmd)
+{
+  unformat_input_t _line_input, *line_input = &_line_input;
+  u8 *app_name = NULL;
+  u32 rule_index = 0;
+  clib_error_t *error = NULL;
+  int rv = 0;
+  int add = 0;
+
+  /* Get a line of input. */
+  if (!unformat_user (input, unformat_line_input, line_input))
+    return error;
+
+  while (unformat_check_input (line_input) != UNFORMAT_END_OF_INPUT)
+    {
+      if (unformat (line_input, "%s rule %u",
+                    &app_name, &rule_index))
+        {
+          if (unformat (line_input, "del"))
+            {
+              add = 0;
+              break;
+            }
+          else
+            {
```

Comment:
either loose the ```else``` block or explicitly test for an ```add``` like here https://github.com/travelping/vpp/blob/feature/upf/src/plugins/upf/upf.c#L210